[AI tutorials] - train and export models to ONNX (#5138)
ovh/docs: pages/public_cloud/ai_machine_learning/training_tuto_09_train_model_export_onnx/guide.en-gb.md

---
title: AI Training - Tutorial - Train a PyTorch model and export it to ONNX
excerpt: How to train and export an image classification model to ONNX using AI Training
updated: 2023-09-06
---

## Objective

The aim of this tutorial is to show you how to train a custom PyTorch model and export it into **ONNX** ([Open Neural Network Exchange](https://onnx.ai/)) format.

The goal is to train your own image classification model on the famous [MNIST](http://yann.lecun.com/exdb/mnist/) dataset. At the end of the model training, it will be is saved in PyTorch format (`.pt`) and then transformed into ONNX.

Exporting your model in ONNX format allows you to **optimize the inference** of a Machine Learning model.

## Requirements

- Access to the [OVHcloud Control Panel](https://www.ovh.com/auth/?action=gotomanager&from=https://www.ovh.co.uk/&ovhSubsidiary=GB)
- A Public Cloud project created
- The ovhai CLI interface installed on your system (more information [here](/pages/public_cloud/ai_machine_learning/cli_10_howto_install_cli))
- [Docker](https://www.docker.com/get-started) installed and configured to build images.

> - An OCI / Docker image registry. You can use a public registry (such as [Docker Hub](https://hub.docker.com/) for example) or a private registry. Refer to the [Creating a private registry](/pages/public_cloud/containers_orchestration/managed_private_registry/creating-a-private-registry) documentation to create a private registry based on Harbor.

Reviewer: What is OCI ?

Author: [Open Container Initiative (OCI)](https://www.padok.fr/en/blog/container-docker-oci#oci)

```docker
FROM python:3.10

WORKDIR /workspace
ADD . /workspace

RUN pip install -r requirements.txt

CMD [ "python3"  , "/workspace/train_image_classification.py"]

RUN chown -R 42420:42420 /workspace
ENV HOME=/workspace
```

#### Build the Docker image from the Dockerfile

Launch the following command from the **Dockerfile** directory to build your application image:

```console
docker build . -t train-cnn-model-export-onnx:latest
```

> [!primary]
>
> The dot `.` argument indicates that your build context (place of the **Dockerfile** and other needed files) is the current directory.
>
> The `-t` argument allows you to choose the identifier to give to your image. Usually image identifiers are composed of a **name** and a **version tag** `<name>:<version>`. For this example we chose **train-cnn-model-export-onnx:latest**.
>

> [!warning]
>
> Please make sure that the docker image you will push in order to run containers using AI products respects the **linux/AMD64** target architecture. You could, for instance, build your image using **buildx** as follows:
>
> `docker buildx build --platform linux/amd64 ...`
>

#### Push the image into the shared registry

> [!warning]
>

> > The shared registry of AI Deploy should only be used for testing purposes. Please consider attaching your own docker registry. More information about this can be found [here](/pages/public_cloud/ai_machine_learning/training_guide_05_howto_add_registry).

Reviewer: This guide is very similar to  (/pages/public_cloud/ai_machine_learning/gi_07_manage_registry)

It was the original version, dedicated to AI Training. It doesn't provide any examples, while the new one (/pages/public_cloud/ai_machine_learning/gi_07_manage_registry) does and is much more complete. 

Maybe we will delete the 05 guide. In this case, it would be better to replace this documentation by the new one ? (I saw you have done it after, line 328).

> > The shared registry of AI Deploy should only be used for testing purposes. Please consider attaching your own docker registry. More information about this can be found [here](/pages/public_cloud/ai_machine_learning/training_guide_05_howto_add_registry).

Author: ```suggestion
> The shared registry of AI Deploy should only be used for testing purposes. Please consider attaching your own docker registry. More information about this can be found [here](/pages/public_cloud/ai_machine_learning/gi_07_manage_registry).
```

> [!primary]
>
> If your images are stored in a private registry, please follow the documentation [Registries - Use & manage your registries](/pages/public_cloud/ai_machine_learning/gi_07_manage_registry) to add your registry.
>

### Launch the AI Training job

You can launch the training job using the UI or the CLI.

#### Create your training job via UI (Control Panel)

If you do not feel comfortable with commands, this way will be more intuitive for you.

First, go to the `OVHcloud Public Cloud` section of the [Control Panel](https://www.ovh.com/auth/?action=gotomanager&from=https://www.ovh.co.uk/&ovhSubsidiary=GB).

Then, select the `AI Training` section (in the AI & Machine Learning category) and create a new job by clicking `AI Training` > `Launch a new job`.

You can create the job that will train your model and export it to ONNX model. Select the *region* and add your *custom docker image* (`<shared-registry-address>/train-cnn-model-export-onnx:latest`).

Then attach your *Object Storage container* `cnn-model-onnx` and precise the mount directory: `/workspace/models`.

> Finally, *configure your job* and choose the at least `1 GPU`.

Reviewer: ```suggestion
Finally, *configure your job* and choose at least `1 GPU`.
```

#### Create your training job via ovhai CLI

> The following command starts the training.

Reviewer: ```suggestion
The following command starts the training:
```
